android-components/components/browser/state/src/test/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddlewareTest.kt

```diff
@@ -0,0 +1,49 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.browser.state.engine.middleware
+
+import mozilla.components.browser.state.action.TranslationsAction
+import mozilla.components.browser.state.selector.findTab
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.TabSessionState
+import mozilla.components.browser.state.state.createTab
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.concept.engine.Engine
+import mozilla.components.support.test.any
+import mozilla.components.support.test.ext.joinBlocking
+import mozilla.components.support.test.rule.MainCoroutineRule
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.mockito.Mockito
+import org.mockito.Mockito.verify
+class TranslationsMiddlewareTest {
+
+    @get:Rule
+    val coroutinesTestRule = MainCoroutineRule()
+    private val scope = coroutinesTestRule.scope
+    private lateinit var engine: Engine
+    private lateinit var tab: TabSessionState
+    private lateinit var store: BrowserStore
+    private fun tabState(): TabSessionState = store.state.findTab(tab.id)!!
+
+    @Before
+    fun setUp() {
+        engine = Mockito.mock()
+        tab = createTab("https://www.mozilla.org")
+        store = BrowserStore(
+            middleware = listOf(TranslationsMiddleware(engine = engine, scope = scope)),
+            initialState = BrowserState(
+                tabs = listOf(tab),
+            ),
+        )
+    }
+
+    @Test
+    fun `WHEN TranslateExpectedAction is dispatched THEN getSupportedTranslationLanguages is called`() {
+        store.dispatch(TranslationsAction.TranslateExpectedAction(tabId = tab.id)).joinBlocking()
+        verify(engine).getSupportedTranslationLanguages(onSuccess = any(), onError = any())
```

Comment:
Should we also verify  that `onSuccess` and `onError` dispatch the appropriate actions?

Reply:
Ohh, thanks good idea! Added two tests to just check for that.

Reply:
I was doing some research, I think the issue is related to the store, we we are using in the middleware is a different from the one we are verifying on, as the middleware [creates its own store ](https://github.com/mozilla-mobile/firefox-android/blob/ee53af63890c9dcf51a254f73cda97d68acf1929/android-components/components/lib/state/src/main/java/mozilla/components/lib/state/internal/ReducerChainBuilder.kt#L38-L65)😅 . I did some changes locally hope it could help :)

``` diff
diff --git a/android-components/components/browser/state/src/test/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddlewareTest.kt b/android-components/components/browser/state/src/test/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddlewareTest.kt
index ff9cbcd780..3f1e544cfb 100644
--- a/android-components/components/browser/state/src/test/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddlewareTest.kt
+++ b/android-components/components/browser/state/src/test/java/mozilla/components/browser/state/engine/middleware/TranslationsMiddlewareTest.kt
@@ -4,6 +4,7 @@
 
 package mozilla.components.browser.state.engine.middleware
 
+import mozilla.components.browser.state.action.BrowserAction
 import mozilla.components.browser.state.action.TranslationsAction
 import mozilla.components.browser.state.state.BrowserState
 import mozilla.components.browser.state.state.TabSessionState
@@ -13,6 +14,7 @@ import mozilla.components.concept.engine.Engine
 import mozilla.components.concept.engine.translate.TranslationError
 import mozilla.components.concept.engine.translate.TranslationOperation
 import mozilla.components.concept.engine.translate.TranslationSupport
+import mozilla.components.lib.state.MiddlewareContext
 import mozilla.components.support.test.any
 import mozilla.components.support.test.argumentCaptor
 import mozilla.components.support.test.ext.joinBlocking
@@ -62,15 +64,28 @@ class TranslationsMiddlewareTest {
     fun `WHEN TranslateExpectedAction is dispatched AND fetching languages is successful THEN onSuccess occurs`() {
         val supportedLanguages = TranslationSupport(null, null)
         val callbackCaptor = argumentCaptor<((TranslationSupport) -> Unit)>()
-        whenever(engine.getSupportedTranslationLanguages(callbackCaptor.capture(), any()))
-            .thenAnswer { callbackCaptor.capture().invoke(supportedLanguages) }
+        val action = TranslationsAction.TranslateExpectedAction(tabId = tab.id)
+        val middlewareContext = mock<MiddlewareContext<BrowserState, BrowserAction>>()
 
-        store.dispatch(TranslationsAction.TranslateExpectedAction(tabId = tab.id)).joinBlocking()
+        whenever(middlewareContext.store).thenReturn(store)
+
+        translationsMiddleware.invoke(middlewareContext, {}, action)
 
         store.waitUntilIdle()
         scope.testScheduler.advanceUntilIdle()
 
         verify(engine).getSupportedTranslationLanguages(onSuccess = callbackCaptor.capture(), onError = any())
+        callbackCaptor.value(supportedLanguages)
+
+        store.waitUntilIdle()
+        scope.testScheduler.advanceUntilIdle()
+
+        verify(store).dispatch(
+            TranslationsAction.TranslateSetLanguagesAction(
+                tabId = tab.id,
+                supportedLanguages = supportedLanguages,
+            ),
+        )
     }
 
     @Test
```

Reply:
Ohh, thank you so much, I didn't realize that! Seems clear now with your changes. I spent forever trying to figure out why it wasn't working! 

Updated, thank you for the research and adding a sample!

Reply:
Always happy to help :)

Reply:
Just wanting to explain all the pushes after review!: 

Hmmm, something is now going on with the more important [onSuccess](https://firefoxci.taskcluster-artifacts.net/A25vwkUGT1-X58QSktF6HA/0/public/reports/tests/testReleaseUnitTest/index.html) test now. 

I think it has something to do with the dispatchers. I can get that error if I take out the waits locally. I'll take a look tomorrow on how to figure out the dispatcher waits for CI.